### ISSUE TITLE
[Presence] Huge spike in CPU usage/Federation traffic approximately every 25 minutes

### ISSUE THREAD

**Issue by realtyem** [matrix-org/synapse#15878]
There is a timeout for when to send a ping over federation every 25 minutes that keeps a user from being marked 'offline' before the 30 minute timeout hits.

![Federation spike 1](https://github.com/matrix-org/synapse/assets/1582365/2a05bc9d-6510-4df9-af54-b1d96244f028)

This appears to be the replication notifier system ramping up and queueing a bunch of federation sending requests over approximately 1 minute worth of time(give or take a few seconds)
<details><summary>Images</summary>

![Federation spike 2](https://github.com/matrix-org/synapse/assets/1582365/6042747f-3a96-439c-9123-36ad500b1006)

![Federation spike 3](https://github.com/matrix-org/synapse/assets/1582365/260d263f-0bd0-419a-88dc-ea6ae52fa1df)

![Federation spike 4](https://github.com/matrix-org/synapse/assets/1582365/7ece6833-c9d3-4f26-90b9-78a08b483d3a)

</details>

There is a database hit during this to `get_current_hosts_in_room()`, I'm not personally convinced it's contributing to the seriousness of this situation(but included here for completeness).
<details><summary>Images:</summary>

![Federation spike 5](https://github.com/matrix-org/synapse/assets/1582365/2cdee10a-590b-49a9-a7a8-bab73a48f439)

</details>

**Comment by H-Shay** [comment]
I think this is essentially a dupe of https://github.com/matrix-org/synapse/issues/9478 - I am going to close this in favor of that.

**Comment by realtyem** [comment]
I postulate that this is different and more specific(ignoring that #9478 is actually a meta issue similar to where this is housed). I believe that this specific behavior is actually caused by the prefill of presence that is setup on startup of Synapse. 

Look here at part of the `__init__()` function where the wheel timers that decided when things that need to be timed out are initially set. In the case of federation pings, this is set to 25 minutes. Specifially, line 683 is where the relevant timer is calculated(we will ignore the other federation timer that is set, as it is for remote users and we are not responsible for sending presence for them).
https://github.com/matrix-org/synapse/blob/b07b14b494ae1dd564b4c44f844c9a9545b3d08a/synapse/handlers/presence.py#L668-L691

Slightly further down(same function) is where a looping call is setup to run every 5 seconds after waiting 30 seconds(so it doesn't influence the rest of Synapse starting up).:
https://github.com/matrix-org/synapse/blob/b07b14b494ae1dd564b4c44f844c9a9545b3d08a/synapse/handlers/presence.py#L723-L735

After all this initially runs, all the timers are based on this timer, then are reset when timeouts are handled(in this instance and for now) in about 25 minutes, leading to a perpetual repeat of this situation.

I speculate that adjusting the initial set of this timer to spread it over the 25 minute interval(evenly or randomly) would break this load up and prevent the spikes. Subsequent timers will then be also spread out.

**Comment by H-Shay** [comment]
Fair enough! Thanks for adding an explanation, that makes what you are getting at much clearer - I will re-open.

**Comment by realtyem** [comment]
These images/screenshots of metrics are for a single user(for context).

So, what happens when the scale of local users climbs to 100's of thousands? In theory, an exponential increase in traffic.

I'm envisioning something similar to a deduplicating bucket like system with an arbitrary timer(say 5 or 10 seconds) that can accumulate presence data for multiple users(or update it if some change came along in that 5 seconds) to 'append' to outgoing federation traffic. Then, if another need to send data over federation comes in, it can just check that bucket and send it. If no other request has come in within that arbitrary timer, then go ahead and send the presence data by itself. Or some such.

I'm open to clever ideas, if anyone would like to suggest something.
